lib/pp_diff.mli

```diff
@@ -0,0 +1,46 @@
+(************************************************************************)
+(*         *   The Coq Proof Assistant / The Coq Development Team       *)
+(*  v      *   INRIA, CNRS and contributors - Copyright 1999-2018       *)
+(* <O___,, *       (see CREDITS file for the list of authors)           *)
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+(** Computes the diff between two Pp.t structures and return
+versions of each with diffs highlighted as (old, new) *)
+val diff_pp : Pp.t -> Pp.t -> Pp.t * Pp.t
+
+(** Computes the diff between two Pp.t structures and return
+a highlighted Pp.t.  If [show_removed] is true, show separate lines for
+removals and additions, otherwise only show additions *)
+val diff_pp_combined : ?show_removed:bool -> Pp.t -> Pp.t -> Pp.t
+
+(** Raised if the diff fails, e.g. the input can't be lexed *)
+exception Diff_Failure of string
```

Comment:
Turn this into a return type of `diff_pp`.

Reply:
Diff_Failure is raised in 3 places only for "impossible" conditions that would be bugs.  The first can't occur unless pp_diff.ml is incorrectly modified.  The other two can happen only if a caller doesn't follow the condition "If the inputs are not acceptable to the lexer, break the strings into
lists of tokens and call diff_strs, then add_diff_tags with a Pp.t that matches the input lists of strings" (line 22).  I think these bug-type exceptions aren't worth making part of the return type.

I've noticed that library functions such as in String are pretty free about raising exceptions (e.g. String.index may raise Not_found).  Are you philosophically against this design pattern?

On the other hand, there are some strings won't lex, which will raise an exception (Proof_diffs:tokenize_string).  I want to tweak the lexer so it has a mode in which it can lex any string to eliminate this.  I plan to do this soon in a separate PR rather than add still more to this one.

Reply:
Exceptions are indeed a double-edged sword.

There are different types of exceptions.  In the 3 places where I explicitly raise Diff_Failure, if they ever occur, they represent bugs that should be reported and fixed.  (The other type of exception is a just an alternate way for the caller to receive some info--not a bug in itself at all.  But if the caller fails to check for and handle this type of exception, that is indeed an error by the caller--the case you mention.)

Mine could arguably be asserts, but that seems pretty user-unfriendly.  For my calls to diff, the user can turn off diffs and continue working with no disruption.  In an abundance of caution, I do check for Diff_Failure in my calls and if it ever occurs, I print the context without diffs just in case I missed something.  For now.

I'll add some more explanation in the doc.

Reply:
I think that what you are trying to do is a general practice that is covered in Coq by the `anomaly` function. Probably better to reuse this as this will make the intent clearer.

Reply:
OK.  Is it legitimate to add the Anomaly exception to CErrors.mli and then filter on it in a try ...  with?

Reply:
No. I may have misunderstood your use case. I had only read your comment and not the code. `anomaly` is only to be used for fatal failures. So don't take my comment into account then. Just document what exceptions functions can raised like recommended by Emilio.